File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.py
import sys
from copy import deepcopy

input = sys.stdin.readline

N = int(input())
board = []
for _ in range (N):
    board.append(list(map(int,input().split())))
      
           
def up(board):
    for j in range(N):
        pointer = 0
        for i in range(1, N):
            # 현재 값이 0이 아닐때
            if board[i][j]:
                tmp = board[i][j]
                board[i][j] = 0
                # 포인터가 0을 가리키는 경우
                if board[pointer][j] == 0:
                    board[pointer][j] = tmp
                # 포인터와 현재 값이 같은 경우
                elif board[pointer][j] == tmp:
                    board[pointer][j] *= 2
                    pointer += 1
                # 포인터와 현재 값이 다른 경우
                else :
                    pointer += 1
                    board[pointer][j] = tmp
    return board
def down(board):
    for j in range(N):
        pointer = N - 1
        for i in range(N - 2, -1, -1):
            if board[i][j]:
                tmp = board[i][j]
                board[i][j] = 0
                if board[pointer][j] == 0:
                    board[pointer][j] = tmp
                elif board[pointer][j]  == tmp:
                    board[pointer][j] *= 2
                    pointer -= 1
                else:
                    pointer -= 1
                    board[pointer][j] = tmp
    return board

def left(board):
    for i in range(N):
        pointer = 0
        for j in range(1,N):
            if board[i][j]:
                tmp = board[i][j]
                board[i][j] = 0
                if board[i][pointer] == 0:
                    board[i][pointer] = tmp
                elif board[i][pointer] == tmp:
                    board[i][pointer] *= 2
                    pointer += 1
                else :
                    pointer +=1
                    board[i][pointer] = tmp
    return board

def right(board):
    for i in range(N):
        pointer = N - 1
        for j in range(N - 2, -1, -1):
            if board[i][j]:
                tmp = board[i][j]
                board[i][j] = 0
                if board[i][pointer] == 0:
                    board[i][pointer] = tmp
                elif board[i][pointer]  == tmp:
                    board[i][pointer] *= 2
                    pointer -= 1
                else:
                    pointer -= 1
                    board[i][pointer] = tmp
    return board

def dfs(board, cnt):
    if cnt == 5:
        # 2차원 배열 요소 중 가장 큰 값 반환
        return max(map(max, board))

    # 상, 하, 좌, 우로 움직여 리턴한 값들 중 가장 큰 값 반환
    # board를 꼭 deepcopy하여 함수를 거친 board값이 다음 함수에 들어가지 못하도록 해주어야 한다.
    return max(dfs(up(deepcopy(board)), cnt + 1), dfs(down(deepcopy(board)), cnt + 1), dfs(left(deepcopy(board)), cnt + 1), dfs(right(deepcopy(board)), cnt + 1))

print(dfs(board, 0))

File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    static int N;
    static int[][] board;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        board = new int[N][N];

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        System.out.println(dfs(board, 0));
    }

    static int dfs(int[][] board, int cnt) {
        if (cnt == 5) {
            int max = 0;
            for (int[] row : board) {
                for (int value : row) {
                    max = Math.max(max, value);
                }
            }
            return max;
        }

        return Math.max(
            Math.max(dfs(up(copy(board)), cnt + 1), dfs(down(copy(board)), cnt + 1)),
            Math.max(dfs(left(copy(board)), cnt + 1), dfs(right(copy(board)), cnt + 1))
        );
    }

    static int[][] up(int[][] board) {
        for (int j = 0; j < N; j++) {
            int pointer = 0;
            for (int i = 1; i < N; i++) {
                if (board[i][j] != 0) {
                    int tmp = board[i][j];
                    board[i][j] = 0;

                    if (board[pointer][j] == 0) {
                        board[pointer][j] = tmp;
                    } else if (board[pointer][j] == tmp) {
                        board[pointer][j] *= 2;
                        pointer++;
                    } else {
                        pointer++;
                        board[pointer][j] = tmp;
                    }
                }
            }
        }
        return board;
    }

    static int[][] down(int[][] board) {
        for (int j = 0; j < N; j++) {
            int pointer = N - 1;
            for (int i = N - 2; i >= 0; i--) {
                if (board[i][j] != 0) {
                    int tmp = board[i][j];
                    board[i][j] = 0;

                    if (board[pointer][j] == 0) {
                        board[pointer][j] = tmp;
                    } else if (board[pointer][j] == tmp) {
                        board[pointer][j] *= 2;
                        pointer--;
                    } else {
                        pointer--;
                        board[pointer][j] = tmp;
                    }
                }
            }
        }
        return board;
    }

    static int[][] left(int[][] board) {
        for (int i = 0; i < N; i++) {
            int pointer = 0;
            for (int j = 1; j < N; j++) {
                if (board[i][j] != 0) {
                    int tmp = board[i][j];
                    board[i][j] = 0;

                    if (board[i][pointer] == 0) {
                        board[i][pointer] = tmp;
                    } else if (board[i][pointer] == tmp) {
                        board[i][pointer] *= 2;
                        pointer++;
                    } else {
                        pointer++;
                        board[i][pointer] = tmp;
                    }
                }
            }
        }
        return board;
    }

    static int[][] right(int[][] board) {
        for (int i = 0; i < N; i++) {
            int pointer = N - 1;
            for (int j = N - 2; j >= 0; j--) {
                if (board[i][j] != 0) {
                    int tmp = board[i][j];
                    board[i][j] = 0;

                    if (board[i][pointer] == 0) {
                        board[i][pointer] = tmp;
                    } else if (board[i][pointer] == tmp) {
                        board[i][pointer] *= 2;
                        pointer--;
                    } else {
                        pointer--;
                        board[i][pointer] = tmp;
                    }
                }
            }
        }
        return board;
    }

    static int[][] copy(int[][] original) {
        int[][] newBoard = new int[N][N];
        for (int i = 0; i < N; i++) {
            System.arraycopy(original[i], 0, newBoard[i], 0, N);
        }
        return newBoard;
    }
}